Support hot restart for DDC library bundle format

srujzs opened the pull request:
https://github.com/dart-lang/webdev/issues/2516
https://github.com/dart-lang/webdev/issues/2561

- Adds logic to emit newly compiled sources to a file that can be read by the bootstrapper.
- Adds bootstrapping logic to reload scripts as needed. This involves implementing the necessary hot restart callback, fetching and processing the emitted file of newly compiled sources, cache busting, and loading the scripts onto the page. A lot of this logic is similar or identical to what we have for internal hot restart support.
- Runs existing hot restart tests with the new bundle format.
- Adds meta tag to run with utf-8 like in https://github.com/flutter/flutter/pull/161493.
- Uses DWDS 24.3.3.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

srujzs committed:
Support hot restart for DDC library bundle format

- Adds logic to emit newly compiled sources to a file that can
be read by the bootstrapper.
- Adds bootstrapping logic to reload scripts as needed. This
involves implementing the necessary hot restart callback, fetching
and processing the emitted file of newly compiled sources, cache
busting, and loading the scripts onto the page. A lot of this
logic is similar or identical to what we have for internal hot
restart support.
- Runs existing hot restart tests with the new bundle format.
- Adds meta tag to run with utf-8 like in
https://github.com/flutter/flutter/pull/161493.
- Uses DWDS 24.3.3.

srujzs committed:
Merge remote-tracking branch 'upstream/main' into hotrestart

srujzs committed:
Update pubspec checksum

nshahan commented on packages/flutter_tools/lib/src/web/bootstrap.dart:
```diff
@@ -196,6 +209,8 @@ let _scriptUrls = {
     ];
 
     let loadConfig = new window.\$dartLoader.LoadConfiguration();
+    // TODO(srujzs): Verify this is sufficient for Windows.
```
Question for @bkonyi (maybe?): Do the tests here run on various platforms in the Flutter CI? Windows?

srujzs replied:
Well, it turns out `forbidden_imports_test.dart` makes sure I don't import `dart:io`, so maybe I shouldn't go down that route. :) I'll keep a todo here to make this work for Windows if needed for now.

srujzs replied:
Hmm, it looks like I should access `isWindows` using some of the abstractions we have in Flutter instead. Updated to use that.

bkonyi replied:
We definitely run the native targets using various platforms, but I'm not sure about web. Maybe @christopherfujino would know off the top of his head?

christopherfujino replied:
not sure off the top of my head.

nshahan commented on packages/flutter_tools/lib/src/web/bootstrap.dart:
```diff
@@ -216,11 +231,82 @@ let _scriptUrls = {
+        if (url == 'dart_sdk.js') {
+          return dartDevEmbedder.debugger.getSourceMap('dart_sdk');
+        }
+        url = url.replace(".lib.js", "");
```
Where is the coupling here? The ".lib.js" in the file name is configured somewhere in the compile actions?

srujzs replied:
Alas, the frontend server hard-codes it in: https://github.com/dart-lang/sdk/blob/ba9ef09163b97992dacc3a18b5f8846958818cf8/pkg/frontend_server/lib/src/javascript_bundle.dart#L270, so there's an implicit dependency there. I don't think there's a way in the recompile to tell it what to use.

nshahan commented on packages/flutter_tools/lib/src/web/bootstrap.dart:
```diff
@@ -216,11 +231,82 @@ let _scriptUrls = {
+              numToLoad++;
+
+              var el = document.getElementById(script.id);
+              if (el) el.remove();
```
Seems like this is removing a script from the page but do we know for sure that it is no longer needed? I wonder what happens after a hot reload if you step into old code from a previous generation that is still kept alive through some existing state in the app but the script is gone.

srujzs replied:
Hot restart reinitializes the state so I don't think such a scenario is possible unless you go through interop, but we don't really plan on handling that scenario.

nshahan commented on packages/flutter_tools/lib/src/web/bootstrap.dart:
```diff
@@ -2,6 +2,8 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'dart:io';
```
Sorry if I missed it but is this new import used?

srujzs replied:
No worries: `Platform.isWindows`.

nshahan replied:
Ahh of course. I dislike the syntax highlighting of strings in this review UI. It's hard quickly to see the interpolation bits.

srujzs committed:
Merge remote-tracking branch 'upstream/main' into hotrestart

srujzs committed:
Amend bootstrap_test and move restart test to utils

The bootstrap test was testing for unused code that was deleted.
The hot restart tests were timing out so they need to be forked.

srujzs committed:
Move additionalCommandArgs to _testProject instead

srujzs committed:
Merge remote-tracking branch 'upstream/main' into hotrestart

srujzs committed:
Remove dart:io import

srujzs committed:
Use globals.platform.isWindows to determine if Windows

bkonyi commented on packages/flutter_tools/lib/src/isolated/devfs_web.dart:
```diff
@@ -146,9 +146,22 @@ class WebAssetServer implements AssetReader {
+  /// If [writeRestartScripts] is true, writes a list of sources mapped to their
+  /// ids to the file system that can then be consumed by the hot restart
+  /// callback.
+  // TODO(srujzs): We should probably include source maps here to reload as well
```
Should we have an issue associated with this TODO?

srujzs replied:
This is actually a stale TODO, so I'm removing it. I didn't realize we're already storing the new source maps in the filesystem on a recompile. The browser fetches the map based on the filename, so there's nothing for us to do here.

bkonyi commented on packages/flutter_tools/lib/src/isolated/devfs_web.dart:
```diff
@@ -158,7 +171,19 @@ class WebAssetServer implements AssetReader {
       final String path = moduleName.replaceAll('.js', '');
       _modules[name] = path;
       _digests[name] = _webMemoryFS.files[moduleName].hashCode.toString();
+      srcIds[moduleName] = moduleName;
```
`srcIds[src]` is always `src`. Can `srcIds` be a set instead?

srujzs replied:
Indeed, done!

bkonyi reviewed (approved):
LGTM overall!

srujzs committed:
Add isWindows to various tests, update copyright, remove TODO, change map to set

srujzs committed:
Merge remote-tracking branch 'upstream/main' into hotrestart

srujzs committed:
Remove test driver annotation on utils

srujzs committed:
Merge remote-tracking branch 'upstream/main' into hotrestart

srujzs committed:
Remove unnecessary set

srujzs committed:
Merge remote-tracking branch 'upstream/main' into hotrestart

srujzs committed:
Move utils to test_data to silence analysis
